ccm_web/server/src/lti/lti.service.ts

```diff
@@ -0,0 +1,75 @@
+import type { Express, Request, Response } from 'express'
+import { BeforeApplicationShutdown, Injectable } from '@nestjs/common'
+import { ConfigService } from '@nestjs/config'
+import { IdToken, Provider as LTIProvider } from 'ltijs'
+import Database from 'ltijs-sequelize'
+
+import baseLogger from '../logger'
+import { DatabaseConfig, LTIConfig } from '../config'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+// ltijs docs: https://cvmcosta.me/ltijs/#/
+@Injectable()
+export class LTIService implements BeforeApplicationShutdown {
+  provider: LTIProvider | undefined
```

Comment:
LTI Service and LTI Middleware seems to be more or less have a Inheritance pattern. LTI middleware you seems to define the function and LTI Service is filling in the details of it. Can you direct me to a documentation about your approach? I don't understand it fully why are you doing it this way.

Reply:
Hm, okay. I still think you and @lsloan had a question about whether we could coalesce `LTIService` and `LTIMiddleware`. Short answer is I don't know. We need to run an `async` method on instance creation, and I'm not sure if this can be done with middleware. Most of what I found online suggests using a dynamic service (through a factory method).  See https://stackoverflow.com/questions/56239773/nestjs-how-to-configure-middleware-with-async-await This is essentially what I'm doing now in `lti.module.ts` with the custom provider there.

There is that other project from Texas that uses Nest and `litjs`, and they are using `onModuleInit` (another lifecycle hook) to set up the instance when the module is loaded. However, when I tried that, I couldn't also use the `ConfigService` (which I need to set up the `ltijs` `Provider`. https://github.com/txstate-etc/attendance-node/blob/main/api/src/lti/lti.middleware.ts

I think the current arrangement seems to work, so my inclination would be not to change it too much -- unless there is some shortcoming of this I'm not seeing.